Add support for shareable modules

thoov opened the pull request:
Enables the config of `scripty.modules` which specifies an array of modules to include in the search space of script executables.

Implements #68

thoov commented:
cc: @stefanpenner

thoov commented:
@searls Could you review this? I am open to any feedback / changes you come up with.

searls commented on lib/load-option.js:
```diff
@@ -12,4 +12,14 @@ module.exports = function loadOption (name) {
   if (pkg === 'false') return false
   if (pkg) return pkg
+
+  if (process.env['npm_package_scripty_' + name + '_0']) {
```
hah, TIL how npm handles array types in its env var unrolling of nested package.json props.

This method was already a little crusty before this addition, but its intent is definitely too mysterious now. I'd recommend yanking out a few private functions at least that reveal intent first, and then adding this.


If this file had previously looked like the following listing, how might you name & factor this new behavior?

```js
var _ = require('lodash')

module.exports = function loadOption (name) {
  if (envVarSet(posixEnvVarName(name))) {
    return boolEnvVarValue(posixEnvVarName(name))
  } else if (envVarSet(packageEnvVarName(name))) {
    return boolEnvVarValue(packageEnvVarName(name))
  }
}

function boolEnvVarValue (envVarName) {
  var value = process.env[envVarName]

  if (value === 'true') {
    return true
  } else if (value === 'false') {
    return false
  } else {
    return value
  }
}

function envVarSet (envVarName) {
  return !!process.env[envVarName]
}

function posixEnvVarName (optionName) {
  return 'SCRIPTY_' + _.snakeCase(optionName).toUpperCase()
}

function packageEnvVarName (optionName) {
  return 'npm_package_scripty_' + optionName
}
```

thoov replied:
I have updated my PR with this change plus my new ones on top of that

searls commented on lib/resolve-script/glob-patterns.js:
```diff
@@ -1,14 +1,31 @@
+        globArray.push(path.join(moduleLocation, '*'))
+      }
+    })
+  }
```
This method was pretty tidy previously and now has a 2-nested branch condition shoehorned into it. Could we extract this to a well named private function? Maybe:

```js
return [
  stuff,
  that, 
  was, 
  here
].concat(arrayOfNewNestedStuff(options.modules))
```

searls commented on package.json:
```diff
@@ -48,7 +48,8 @@
     "glob": "^7.0.3",
-    "lodash": "^4.8.2"
+    "lodash": "^4.8.2",
+    "resolve-pkg": "^1.0.0"
```
If we merge `resolve-pkg`, I think I get a free coffee for having 10 different dependencies that all implement the resolve algorithm.

Off the top of my head, [quibble relies on substack/browserify `resolve`](https://github.com/testdouble/quibble/blob/master/package.json#L22) whereas [testdouble uses the built-in `require.resolve`](https://github.com/testdouble/testdouble.js/blob/1bbf695953912d9c2925bd8c570c9ff0c3328dbf/src/replace/module/require-actual.js#L22). Is there any chance this change could rely on one of those two instead?

thoov replied:
The issue with `require.resolve` is that if your module does not contain a "main" property it will not resolve. So when I was testing locally with our internal module it would not resolve because of this. We could force a "blank" main file but that seems a little weird (and out of scope of this). Also the result of `require.resolve` is the entry path (package root + main). This means we would have to do some processing to "strip out" these extra paths. The same issue exists in `resolve`.

resolve-pkg's tag line is: 
> Resolve the path of a package regardless of it having an entry point

I agree that bringing in new packages should be avoided at all costs. However, it seems like the 3 options are to either bring this in, write a bunch of custom logic to do the resolve, or force a convention on the modules. My option is that the first options seems like the best fit. I am open to any counter points?

searls replied:
Welp, ok then.

searls commented on yarn.lock:
```diff
@@ -0,0 +1,1906 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
scripty uses npm and an npm package-lock.json. Could you please use npm & update its package-lock? I'm uncomfortable with the idea of having two (potentially divergent) sources of truth as to the library's nested dependencies.

thoov replied:
Oops 😄 (total just assumed yarn... I wonder if there is a wrapper that would just delegate to either yarn or npm based on the presence of a yarn.lock or package-lock 🤔). I will get rid of this and use `package-lock.json`.

thoov commented on README.md:
```diff
@@ -213,6 +213,42 @@ You can configure either or both of `"path"` and `"windowsPath"` to custom
+take priority over ones defined in modules. The resolution of `bar` would resolve to
+`root.node_modules.packageA.scripts.bar` as packageA was the first module defined
+in the `scripty.modules` config.
+
```
👍

thoov commented:
@searls Thanks for the review! I believe I have addressed all of your comments (expect the windows one which I left a clarifying comment). Let me know if my response / changes make sense or if you have any other comments / suggestions.

searls commented on lib/resolve-script/glob-patterns.js:
```diff
@@ -1,14 +1,31 @@
+
+  if (options && Array.isArray(options.modules)) {
+    options.modules.forEach(function (moduleName) {
+      const modulePath = resolvePkg(moduleName + '/scripts')
```
I should be more clear:

1. Windows command support is implemented in a custom way in Scripty. If you're on Windows, it will first look in a mirrored directory hierarchy under `scripts-win/`, not `scripts/`. That means for you to keep Windows support working consistently, this feature would need to resolve windows paths in the same way other scripts are loaded (via this [bit of code here](https://github.com/testdouble/scripty/blob/a003adc28b39c57bffa394714d5da6e2bca69d06/lib/resolve-script/script-dirs.js#L13-L17)). 

2. Whoops, fixed the Appveyor permission error. You can see a [build log here](https://ci.appveyor.com/project/searls/scripty/build/job/1k6wkfw5a4owk8ff). What I think this PR really needs to be sure it works is a "safe" (i.e. integration) test that uses a new set of minimal nested module fixtures to verify that script-module loading works as you expect it to, and that also exercises scripts-win. I think if you look under the `safe/` path you should find readily imitatible tests to start from

thoov commented:
@searls I added a "safe" test that exercises the win32 flow (thanks for the clarification and pointing me to those examples). I believe this to be in a good spot now. Could you review (glob-patterns.js and script-dirs.js) as those were the places that I changed the most between your last review and now.

searls commented:
Landed in v.1.8.0! 🎊
